File: rp-config-common/src/main/java/rainpoetry/config/commconf/common/ByteUnits.java
package rainpoetry.config.commconf.common;


import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author RainPoetry
 */

public class ByteUnits {

	private static final String pattern = "(-?[0-9]+)([a-z]+)?";

	private static final Map<String, ByteUnit> byteSuffixes;

	static {
		byteSuffixes = Maps.newHashMap()
				.pairs("b", ByteUnit.BYTE)
				.pairs("k", ByteUnit.KiB)
				.pairs("kb", ByteUnit.KiB)
				.pairs("m", ByteUnit.MiB)
				.pairs("mb", ByteUnit.MiB)
				.pairs("g", ByteUnit.GiB)
				.pairs("gb", ByteUnit.GiB)
				.pairs("t", ByteUnit.TiB)
				.pairs("tb", ByteUnit.TiB)
				.pairs("p", ByteUnit.PiB)
				.pairs("pb", ByteUnit.PiB)
				.unModify();
	}

	public static long byteStringAS(String format, ByteUnit unit) {
		String lower = format.toLowerCase().trim();
		if (match(lower)) {
			Matcher matcher = Pattern.compile(pattern).matcher(lower);
			if (matcher.find()) {
				String suffix = matcher.group(2);
				long number = Long.parseLong(matcher.group(1));
				if (suffix != null) {
					return unit.convertFrom(number, byteSuffixes.get(suffix));
				} else {
					return unit.convertFrom(number, unit);
				}
			}
		}
		throw new NumberFormatException("Failed to parse time string: " + format);
	}

	public static boolean match(String format) {
		return format.matches(pattern);
	}

	public static long byteStringAsBytes(String str) {
		return byteStringAS(str, ByteUnit.BYTE);
	}

	public static long byteStringAsKb(String str) {
		return byteStringAS(str, ByteUnit.KiB);
	}

	public static long byteStringAsMb(String str){
		return byteStringAS(str, ByteUnit.MiB);
	}

	public static long byteStringAsGb(String str){
		return byteStringAS(str, ByteUnit.GiB);
	}

	public static void main(String[] args) {
		System.out.println(byteStringAS("2M", ByteUnit.GiB));
	}

}

File: config-scala/src/main/scala/rainpoetry/config/scala/tools/ByteUnits.scala
package rainpoetry.config.scala.tools

import java.util.Locale

import java.lang.{Long => JLong}

import scala.collection.immutable

/*
 * author:  RainPoetry
 * date:  2019/8/23
 * description: 
 */

object ByteUnits {

  val pattern = "(-?[0-9]+)([a-z]+)?".r

  val byteSuffixes = immutable.HashMap[String, ByteUnit](
    "b" -> ByteUnit.BYTE,
    "k" -> ByteUnit.KiB,
    "kb" -> ByteUnit.KiB,
    "m" -> ByteUnit.MiB,
    "mb" -> ByteUnit.MiB,
    "g" -> ByteUnit.GiB,
    "gb" -> ByteUnit.GiB,
    "t" -> ByteUnit.TiB,
    "tb" -> ByteUnit.TiB,
    "p" -> ByteUnit.PiB,
    "pb" -> ByteUnit.PiB
  )

  def byteStringAs(str: String, unit: ByteUnit): Long = {
    val lower = str.toLowerCase(Locale.ROOT).trim
    lower match {
      case pattern(number, suffix) =>
        val value = JLong.parseLong(number)
        if (suffix != null && !byteSuffixes.contains(suffix)) {
          throw new NumberFormatException("Invalid suffix: \"" + suffix + "\"");
        }
        if (suffix != null) {
          unit.convertFrom(value, byteSuffixes(suffix))
        } else {
          unit.convertFrom(value, unit)
        }
      case _ => throw new NumberFormatException(s"Failed to parse time string: ${str}")
    }
  }

  def byteStringAsBytes(str: String): Long = {
    byteStringAs(str, ByteUnit.BYTE)
  }

  def byteStringAsKb(str: String): Long = {
    byteStringAs(str, ByteUnit.KiB)
  }

  def byteStringAsMb(str: String): Long = {
    byteStringAs(str, ByteUnit.MiB)
  }

  def byteStringAsGb(str: String): Long = {
    byteStringAs(str, ByteUnit.GiB)
  }
}
